scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -0,0 +1,140 @@
+(ns cook.kubernetes.controller
+  (:require [cook.kubernetes.api :as api]
+            [clojure.tools.logging :as log])
+  (:import (io.kubernetes.client.models V1Pod)))
+
+
+
+
+;
+;   Wire up a store with the results.
+;
+
+(def lock-object (Object.))
+(defn calculate-lock
+  "Given a pod-name, return an object suitable for locking for accessing it."
+  [_]
+  ; TODO: Should do lock sharding based on hash of pod-name.
+  lock-object)
+
+(defn expected-state-equivalent?
+  "Is the old and new state equivalent?"
+  [old-state new-state]
+  false ; TODO
+  )
+
+(defn existing-state-equivalent?
+  "Is the old and new state equivalent?"
+  [old-state new-state]
+  false ; TODO
+  )
+
+(defn remove-finalization-if-set
+  [api-client expected-state-dict pod]
+  (api/remove-finalization-if-set api-client pod)
+  expected-state-dict)
+
+(defn kill-task
+  [api-client expected-state-dict pod]
+  (api/kill-task api-client pod)
+  expected-state-dict)
+
+(defn launch-task
+  [api-client expected-state-dict]
+  (api/launch-task api-client expected-state-dict)
+  expected-state-dict)
+
+(defn update-or-delete!
+  [map-atom key value]
+  (if (nil? value)
+    (swap! map-atom dissoc map key)
+    (swap! update map-atom key value)))
+
+;; REVIEW: Review handle-status-update
+(defn pod-has-just-completed
```

Comment:
`pod-completed`

I thought the whole idea was to not worry about "just" anymore :)

Reply:
As soon as we enter this state, the next step is to going to be updating datomic and immediately transitioning out of this :existing/running state.